### ISSUE TITLE
Create a regular expression to identify citations

### ISSUE THREAD

**Issue by waldoj** [openlawdc/dc-decoded#4]
Identifying and linking cross-references requires being able to identify them within the text of the laws. Write a PCRE to accomplish that. (That gets dropped into The State Decoded's configuration file.) Here are [some example PCREs for state codes](https://github.com/statedecoded/law-identifier).

**Comment by tmcw** [comment]
I've been using `/^§ (\d+)\-(\d+)\.(\d+)?/` so far.

**Comment by waldoj** [comment]
And I'm sure that does the trick the bulk of the time. I'm [working on a more rigorous PCRE at the moment](https://github.com/statedecoded/law-identifier/blob/master/Washington-DC.md), but once I pare it down, it's possible I'll wind up right where you started. :)

**Comment by waldoj** [comment]
Yeah, you're getting some bad matches, e.g. the subsubsection here:

```
{
    "title": "§ 24-221.01c. Credits for good behavior, rehabilitation programs, work details, and special projects.",
    "section": "24",
    "subsection": "221",
    "subsubsection": "01"
},
```

The first portion of the section identifier is the title, so rather than calling 24 the "section," you can identify it as the "title." I don't know that there's any benefit to breaking up the section identifier in any more detail than that (i.e., subsection, subsubsection). The only use for that, I suppose, is for sorting, but I think it might be more useful to provide a `sort_by` attribute. "221" and "01c," on their own, don't really indicate anything.

**Comment by waldoj** [comment]
Also, `DC_CODE_Title 29A.json` is only 2 bytes long. It's the "A" that's throwing things off there, I assume.

**Comment by tmcw** [comment]
Working on more of these [in parse.js](https://github.com/openlawdc/dc-decoded/blob/master/parse.js#L5-L22). Btw, I'm idling in #openlawdc on IRC if you want to go over any specifics

**Comment by waldoj** [comment]
I'd join you, but it's my wife's birthday, and I don't want to try her patience. :)

**Comment by waldoj** [comment]
OK, [got it](https://github.com/statedecoded/law-identifier/blob/master/Washington-DC.md):

```
((§\s)?)([0-9]{1,2})([A-Z]?)-([0-9]{3,4})((((\.)([0-9]{2}))?)([a-z]?))
```

It needs more real-world testing, but it's a good start. Now that I've got the whole code as JSON, I'll take the `title` element from a few thousand laws, hack off everything after the first `.`, try the PCRE against those, and manually audit the results. I fear that there are some colon-bearing components that remain to be discovered. :-/

**Comment by tmcw** [comment]
Cool, are you working on a larger parser somewhere?

**Comment by waldoj** [comment]
I'm not...other than The State Decoded itself. :)

**Comment by waldoj** [comment]
OK, I've read through several thousand section identifiers, and I feel good about this PCRE. Specifically, I was looking for, but did not find:
- trailing alpha identifiers (e.g., the "a" in "50-1501.03a") that follow "z"
- trailing decimal-separated identifiers (e.g., the "03" in "50-1501.03") that are less than or greater than two digits
- alternating decimal sort methods (e.g., "10.1, 10.2, [...], 10.9, 10.10" vs. "10.10, 10.11, 10.12, [...], 10.20, 10.21")

I did find that frequently trailing decimal-separated identifiers are not sequential, e.g. .01, .02, .50, .51. ,52. That presents no obstacle to The State Decoded, but I mention it in case it is problematic for you.

**Comment by konklone** [comment]
A note - in the example above, `§ 24-221.01c`, that does not cite any subsections. If you look at the [subchapter above it](http://law.justia.com/codes/district-of-columbia/2012/division-iv/title-24/chapter-2/subchapter-iii/), you can see that it's how they insert new sections between others that already existed. So it's a peer to simpler-looking sections, like `§ 24-271`. Now [§ 24-221.01c](http://law.justia.com/codes/district-of-columbia/2012/division-iv/title-24/chapter-2/subchapter-iii/section-24-221-01c.html) does have potentially citable `(a)`, `(b)`, `(c)`, and `(d)` subsections, but I'm not sure how that would appear in the wild. 

This happens in the US Code all the time too, so top-level section identifiers can get real tortuous. The [USC regex here](https://github.com/unitedstates/citation/blob/master/citations/usc.js#L17) handles a lot of different cases like this, including ranges. I'm not sure how likely the DC Code cites are to get like that - but I did notice when poring through the DC Code that it actually cites the US Code quite a bit (probably moreso than other states, given DC's unique federal oversight).

**Comment by waldoj** [comment]
> A note - in the example above, § 24-221.01c, that does not cite any subsections.

I was just using Tom's nomenclature. :)

> Now [§ 24-221.01c](http://law.justia.com/codes/district-of-columbia/2012/division-iv/title-24/chapter-2/subchapter-iii/section-24-221-01c.html) does have potentially citable `(a)`, `(b)`, `(c)`, and `(d)` subsections, but I'm not sure how that would appear in the wild.

The definition of "the wild" varies—cross references are different than, say, court decisions. In the D.C. court decisions that I surveyed, the citations for subsections are always parenthetical, e.g. "D.C. Code sections 1-2272(8) and (18)" or "D.C. Code §47-1002(8)". Many state codes use this solution of appending a letter at the end to insert new sections at a later date, and the practice of wrapping the first subsection identifier in parentheses seems to be near-universal.

**Comment by tmcw** [comment]
Hm, so one cannot go from citation id -> categorization, because `§ 24-221` could be title 24, chapter 2, section 21 - or it could be title 24, chapter 22, section 1?

**Comment by konklone** [comment]
No, fortunately the citation is unambiguous - `§ 24-221` is title 24, section 221.

Everything besides title and section - divisions, chapters, subchapters, parts, subparts, etc - is not relevant to legal citations. Those extra levels are there for easier organizing in books, and to have a hierarchy where each level is cognitively digestible. This is how the US Code works as well (and this is also discussed a bit in unitedstates/uscode#3).

**Comment by konklone** [comment]
To expand a little bit - you can still, given a citation, figure out which division/chapter/subchapter/part/etc it is in, but it would require having indexed the full spine of the DC Code first, so that when you encounter a legal citation (like `§ 24-221`) you can do a DB lookup (or something) to establish its location in the full hierarchy. But you would need that lookup step, because that information isn't encoded in the citation itself. 

Producing a spine, like described in #9, would give anyone the information they needed to perform that lookup. It's probably helpful to have an easy way to just get at the titles and sections, though, since a lot of tasks don't need the full hierarchy. (That's what unitedstates/uscode#3 is actually about - I added a flag to the parser that flattened the output to just title+section.)

But with the full hierarchy, you can do what @tauberer talked about in that ticket, which is to orient feeds of matching content around categories at whatever level you want - chapters, subchapters, etc. - as long as you're willing to do that lookup step on every matched cite.

**Comment by tmcw** [comment]
Huh, there are some oddities: in Title 1:

```
District of Columbia Official Code 2001 Edition Currentness
Division I. Government of District.
 Title 1. Government Organization. (Refs & Annos)
 Chapter 4. Delegate to the House of Representatives.

DC CODE D. I, T. 1, Ch. 4, Refs & Annos

                Current through December 11, 2012                               

             

END OF DOCUMENT
```

Looks like

> DC CODE D. I, T. 1, Ch. 4, Refs & Annos

Is the ID format for this chunk.

**Comment by konklone** [comment]
Gross. Looks like it's referring just to a chapter, at least, which we may not care about. At least this one has the words "DC CODE" before it to disambiguate.

**Comment by waldoj** [comment]
_sigh_

I think Eric's right about this—these all appear to be cross-references to structural units (chapters, titles, parts, articles, etc.). Those are notoriously difficult to identify and link to, and I recommend regarding those as a different issue than identifying citations of individual laws. Citations to structural units appear to be created on the fly, every time a bill is written, with nobody ever sitting down and deciding what they ought to look like as a standard practice.

**Comment by konklone** [comment]
I agree - unless the DC Code is chock full of these things, and tends to use them in favor of citing individual sections, I think you can feel comfortable ignoring them and detecting cross references only to sections.

**Comment by konklone** [comment]
So Waldo, I didn't end up using your regex directly in my [DC Code parser](https://github.com/unitedstates/citation/blob/master/citations/dc_code.js) - I have a couple of them, and they are more permissive (doesn't check for numbers of digits, for example), and perhaps less rigorous. 

I've also made no effort to tackle ranges, or multiple mentions - for example, there's a commented-out todo at the bottom of [the dc_code test suite](https://github.com/unitedstates/citation/blob/master/test/dc_code.js#L112) that shows a snippet from a Mayoral Order containing "D.C. Official Code § 1-204.22(2) and (11)".

But I think it will work pretty well! And it does handle both inline and absolute references, across a variety of DC legal documents.

**Comment by waldoj** [comment]
When looking for cross-references, having a rigorous regular expression isn't important for 99.9% of laws. For your remaining edge cases, you're going to have things like dollar values or mentions of U.S. Code sections that are erroneously linked, but almost nobody is going to ever see those laws, and something close to zero people will ever click on one of those wrongly linked citations. For the Florida Statutes, I could use `s. ([0-9]+)-([0-9]+)` and, as much as it would pain me, that'd basically do the trick.

It's really when applying that regex to _other_ documents (e.g., court rulings, legislation, AG opinions, that sort of thing) that it becomes important to be rigorous.

**Comment by konklone** [comment]
Mmm. Well, I'd like to do that too - my plan for Sunday is to drag down bills and proposed/final rules from Open States and the DC Register, yank the cites out of them, and then brutally shackle them to the pieces of the DC Code they relate to. I guess I'll just have to see it in action and work from there.

**Comment by waldoj** [comment]
That sounds like a pretty violent Sunday.

**Comment by konklone** [comment]
I'm going to close this as resolved along with openlawdc/code-browser#4, at least for the time being, unless there's a desire to integrate citation extraction at parse-time (here) rather than at render-time (there).
